### ISSUE TITLE
Buidler support

### ISSUE THREAD

**Issue by alcuadrado** [cgewecke/eth-gas-reporter#116]
I'm opening this issue to continue our conversation from #115.

I tried running this reporter with Buidler, and got into these issues:

1. It requires web3, so it would only work in projects using [buidler-web3](https://www.npmjs.com/package/@nomiclabs/buidler-web3). I think this is a reasonable requirement, and I haven't checked yet how difficult it would be to make it more general.

2. It assumes that `artifacts` from truffle is available. I workarounded this by installing [buidler-truffle5](https://www.npmjs.com/package/@nomiclabs/buidler-truffle5).

3. It only works if web3 supports sync calls. Those aren't supported by Buidler, and I think they aren't by web3 1.0.  I saw [this PR](https://github.com/cgewecke/eth-gas-reporter/pull/66), so I commented out [this check](https://github.com/cgewecke/eth-gas-reporter/blob/master/index.js#L14-L18), but I'm getting this error now:

```
pato@pmbp:gas% npx buidler test --show-stack-traces
Compiling...
An unexpected error occurred: Cannot read property 'url' of undefined

TypeError: Cannot read property 'url' of undefined
    at Object.request (/private/tmp/gas/node_modules/eth-gas-reporter/sync.js:57:10)
    at Object.getLatestBlock (/private/tmp/gas/node_modules/eth-gas-reporter/sync.js:21:17)
    at Object.mapMethodsToContracts (/private/tmp/gas/node_modules/eth-gas-reporter/gasStats.js:353:22)
    at Runner.Gas.runner.on (/private/tmp/gas/node_modules/eth-gas-reporter/index.js:143:65)
    at emitNone (events.js:111:20)
    at Runner.emit (events.js:208:7)
    at start (/private/tmp/gas/node_modules/mocha/lib/runner.js:848:10)
    at Runner.run (/private/tmp/gas/node_modules/mocha/lib/runner.js:876:5)
    at Mocha.run (/private/tmp/gas/node_modules/mocha/lib/mocha.js:612:17)
    at Promise (/private/tmp/gas/node_modules/@nomiclabs/buidler/src/builtin-tasks/test.ts:51:13)
    at new Promise (<anonymous>)
    at SimpleTaskDefinition.config_env_1.internalTask.addOptionalVariadicPositionalParam.setAction [as action] (/private/tmp/gas/node_modules/@nomiclabs/buidler/src/builtin-tasks/test.ts:50:24)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:189:7)
```

It expects a `connection` field from `web3`. I have no idea what that is. Do you?

**Comment by cgewecke** [comment]
Ah great! 

> It requires web3 . . . It only works if web3 supports sync calls

The reporter actually makes it's own RPC calls using a module called `syncRequest`. However, it *does* rely on web3's availability in the test context to get the client's protocol/address/port from [the provider's host property](https://github.com/cgewecke/eth-gas-reporter/blob/master/sync.js#L42-L62). With web3 this value is in different locations on the provider object depending on whether http or websockets is being used. 

What is globally available in the test context in buidler? Ideally this would work equally well with ethers.js and we could just search through the possible provider types looking for the necessary info without requiring any special configuration. 

> It assumes that artifacts from truffle is available. 

Yes, exactly. Was thinking it might be possible to abstract that logic out and perhaps publish a buidler plugin for the reporter which knew about other artifact formats. I think we just need:
+ contract name
+ deployment transaction hash (for contracts deployed through a pre-test process like truffle's migrations)

**Comment by cgewecke** [comment]
One big assumption in the above is that client is in a separate process. That might not be the default case at buidler.

**Comment by cgewecke** [comment]
I'd really like to duplicate the truffle integration tests here with buidler as well. Have been wanting to use it for a long time and would be a good chance to understand how it works.

**Comment by alcuadrado** [comment]
> The reporter actually makes it's own RPC calls using a module called [sync-request](https://www.npmjs.com/package/sync-request). However, it _does_ rely on web3's availability in the test context to get the client's protocol/address/port from [the provider's host property](https://github.com/cgewecke/eth-gas-reporter/blob/master/sync.js#L42-L62). With web3 this value is in different locations on the provider object depending on whether http or websockets is being used.

I see. You can get its host with `web3.currentProvider._provider.provider.host` now. It's deeply nested because Buidler [wraps the provider multiple times](https://github.com/nomiclabs/buidler/blob/master/packages/buidler-core/src/internal/core/providers/construction.ts) adding them functionality each time.

A cleaner way of doing that, but Buidler specific, would be `config.networks[buidlerArguments.network].url`.

> What is globally available in the test context in buidler? Ideally this would work equally well with ethers.js and we could just search through the possible provider types looking for the necessary info or connection strategy without requiring any special configuration.

One of the things that Buidler injects globally is an EIP1993-ish provider. This standard isn't finalized yet, so it may change a little in the future. It's available as `ethereum`, because [this other EIP](https://eips.ethereum.org/EIPS/eip-1102) dictates that name, and it's what dapp browsers are doing.

If this reporter could use the provider directly, it would automatically work for ethers.

The main issue is that this provider is async only.

Also, I have plans for an in-process provider, so not sure if the `sync-request` through HTTP would work in the future.

> Yes, exactly. Was thinking it might be possible to abstract that logic out and perhaps publish a buidler plugin for the reporter which knew about other artifact formats. I think we just need:
> 
> * the ability to fetch artifact by contract name
> * abi
> * the 'unlinked_binary' - (needs investigation, not sure what that corresponds to in terms of solc output )
> * deployment transaction hash (for contracts deployed through a pre-test process like truffle's migrations)

The first three should be pretty easy. I have to think more about the last one, as we don't have a concept equivalent to migrations. We are planning to offer some support in `buidler-truffle4` and `buidler-truffle5`, but I guess we may be able to keep those things in the context of a test.

> One big assumption in the above is that client is in a separate process. That might not be the default case at buidler.

Do you mean the mocha process? It's the same process now, but that could be changed. Creating new processes is trivial with Buidler, we already do that in a few places.

> I'd really like to duplicate the truffle integration tests here with buidler as well. Have been wanting to use it for a long time and would be a good chance to understand how it works.

I think this should be fairly easy once we got the above things working.

**Comment by cgewecke** [comment]
> Do you mean the mocha process?

I *think* I mean the client. For example in truffle if you try to supply ganache as a provider directly to main process through truffle's network configurations, web3 0.x's sync calls freeze the execution thread. I believe sync-request would cause the same problem although it sounds like you're anticipating this already with:

> Also, I have plans for an in-process provider, so not sure if the sync-request through HTTP would work in the future.
---------
> we don't have a concept equivalent to migrations

Oh good. The only issue there is whether people consume pre-deployed contracts in their tests and want data about what those deployments cost. If not it doesn't matter.

> I think this should be fairly easy once we got the above things working.

Agree. Have a clear idea of what needs to be configurable now - thanks so much for all your help here.

**Comment by alcuadrado** [comment]
> > Do you mean the mocha process?
> 
> I _think_ I mean the client. For example in truffle if you try to supply ganache as a provider directly to main process through truffle's network configurations, web3 0.x's sync calls freeze the execution thread. I believe sync-request would cause the same problem although it sounds like you're anticipating this already with:

This is interesting. I restricted the provider's config a little to prevent users from passing web3-provider-engine backed providers. I don't think that is stable enough, and I don't like that it works as a service (i.e. you need to start and stop it) without any clear reason. As Buidler also works as a library, there isn't a single place to start/stop it.

> 
> > Also, I have plans for an in-process provider, so not sure if the sync-request through HTTP would work in the future.
> 
> > we don't have a concept equivalent to migrations
> 
> Oh good. The only issue there is whether people consume pre-deployed contracts in their tests and want data about what those deployments cost. If not it doesn't matter.

I see! I'll try to avoid doing that. I guess it was an optimization, but it gets things really confusing. At least for tools developers.

> 
> > I think this should be fairly easy once we got the above things working.
> 
> Agree. Have a clear idea of what needs to be configurable now - thanks so much for all your help here.

Let me know if you need help with anything else :)

**Comment by cgewecke** [comment]
@alcuadrado Hi, small update. I have a working draft of this in #122. About to open a small flurry of issues at buidler for things I ran into while doing this (they're mostly weird truffle things).

Builder is **really really** well designed and well written. Absolutely fantastic work! Congratulations.

Only technical thing I have to report here is that I'm using the provider 'deep reach' approach to get the client url for the time being. Would like to revisit that - probably publishing a simple plugin is the best thing to do.

Do you have a simple example project lying around that *doesn't* use the truffle plugin? It would be helpful for testing for me.

**Comment by alcuadrado** [comment]
> Buidler is **really really** well designed and well written. Absolutely fantastic work! Congratulations.

Thank you! I'm really glad you like it :)

> About to open a small flurry of issues at buidler for things I ran into while doing this (they're mostly weird truffle things).

I really appreciate this. I'll go through these issues tomorrow.

> Only technical thing I have to report here is that I'm using the provider 'deep reach' approach to get the client url for the time being. Would like to revisit that - probably publishing a simple plugin is the best thing to do.

I couldn't find this in #122. Can you point me to it? Maybe I can help you do it in another way.

Still, publishing a small plugin may be the easiest way to integrate it.

> Do you have a simple example project lying around that _doesn't_ use the truffle plugin? It would be helpful for testing for me.

I don't have a real-world example, as most people just use truffle. I'll create a simple project with `buidler-ethers` tomorrow.

**Comment by cgewecke** [comment]
@alcuadrado Ah, the 'deep reach' is in [lib/config.js](https://github.com/cgewecke/eth-gas-reporter/blob/master/lib/config.js#L31-L60) for the moment. 

>  I'll create a simple project with buidler-ethers tomorrow.
Oh nice, thanks.

**Comment by alcuadrado** [comment]
@cgewecke here's the sample project using ethers. It has a super basic contract with a few tests, but it may be enough as a test case. Let me know if it's not.

I also installed typescript in that project, to have it as an example of how to do it, and cause having autocomplete may help you as `buidler-ethers` is not that well documented yet.

> the 'deep reach' is in [lib/config.js](https://github.com/cgewecke/eth-gas-reporter/blob/master/lib/config.js#L30-L59) for the moment.

If `buidler-ethers` is installed, you can do this with `ethers.provider._buidlerProvider.provider.host`. Or you can use `ethereum.provider.host` in any Buidler project.

**Comment by alcuadrado** [comment]
> If `buidler-ethers` is installed, you can do this with `ethers.provider._buidlerProvider.provider.host`. Or you can use `ethereum.provider.host` in any Buidler project.

Actually, no. This is not the right way to do it. Depending on your network config, Buidler wraps the providers N times. So for example, it can end up being `ethereum.provider.provider.provider.host`. 

I think I do something like:

```js
if (typeof buidlerArguments !== "undefined") {
  return config.networks[buidlerArguments.network].url;
}
```

**Comment by cgewecke** [comment]
@alcuadrado Thanks for the ethers example - perfect.

**Comment by cgewecke** [comment]
There's now a published plugin wrapping the reporter at [buidler-gas-reporter](https://github.com/cgewecke/buidler-gas-reporter). Thanks for all your help with this! Will continue to work on it and make improvements but buidler support has begun at least. 

Closing. Published in `0.2.2`.
